src/GraphQL/Internal/Validation.hs

```diff
@@ -1,104 +1,416 @@
+
+-- * Selections
+
+-- $fragmentSpread
+--
+-- The @spread@ type variable is for the type used to "fragment spreads", i.e.
+-- references to fragments. It's a variable because we do multiple traversals
+-- of the selection graph.
+--
+-- The first pass (see 'validateSelection') ensures all the arguments and
+-- directives are valid. This is applied to all selections, including those
+-- that make up fragment definitions (see 'validateFragmentDefinitions'). At
+-- this stage, @spread@ will be 'UnresolvedFragmentSpread'.
+--
+-- Once we have a known-good map of fragment definitions, we can do the next
+-- phase of validation, which checks that references to fragments exist, that
+-- all fragments are used, and that we don't have circular references.
+--
+-- This is encoded as a type variable because we want to provide evidence that
+-- references in fragment spreads can be resolved, and what better way to do
+-- so than including the resolved fragment in the type. Thus, @spread will be
+-- 'FragmentSpread', following this module's convention that unadorned names
+-- imply that everything is valid.
+
+-- | A GraphQL selection.
+data Selection spread
+  = SelectionField (Field spread)
+  | SelectionFragmentSpread spread
+  | SelectionInlineFragment (InlineFragment spread)
+  deriving (Eq, Show)
+
+-- | A field in a selection set, which itself might have children which might
+-- have fragment spreads.
+data Field spread
+  = Field (Maybe Alias) Name Arguments Directives [Selection spread]
+  deriving (Eq, Show)
 
-{-
-enum DogCommand { SIT, DOWN, HEEL }
+-- | A fragment spread that has a valid set of directives, but may or may not
+-- refer to a fragment that actually exists.
+data UnresolvedFragmentSpread
+  = UnresolvedFragmentSpread Name Directives
+  deriving (Eq, Show)
+
+-- | A fragment spread that refers to fragments which are known to exist.
+data FragmentSpread
+  = FragmentSpread Name Directives (FragmentDefinition FragmentSpread)
+  deriving (Eq, Show)
+
+-- | An inline fragment, which itself can contain fragment spreads.
+data InlineFragment spread
+  = InlineFragment TypeCondition Directives [Selection spread]
+  deriving (Eq, Show)
+
+-- | Traverse through every fragment spread in a selection.
+--
+-- The given function @f@ is applied to each fragment spread. The rest of the
+-- selection remains unchanged.
+--
+-- Note that this is essentially a definition of 'Traversable' for
+-- 'Selection'. However, we probably also want to have other kinds of
+-- traversals (e.g. for transforming values), so best not to bless one kind
+-- with a type class.
+traverseFragmentSpreads :: Applicative f => (a -> f b) -> Selection a -> f (Selection b)
+traverseFragmentSpreads f selection =
+  case selection of
+    SelectionField (Field alias name args directives ss) ->
+      SelectionField <$> (Field alias name args directives <$> childSegments ss)
+    SelectionFragmentSpread x -> SelectionFragmentSpread <$> f x
```

Comment:
optional nit: maybe move `SelectionFragmentSpread <$> f x` to next line for regularity with the other lines